pageserver/src/tenant/blob_io.rs

```diff
@@ -222,19 +224,29 @@ impl<const BUFFERED: bool> BlobWriter<BUFFERED> {
     }
 
     /// Internal, possibly buffered, write function
-    async fn write_all<B: BoundedBuf<Buf = Buf>, Buf: IoBuf + Send>(
+    async fn write_all<Buf: IoBuf + Send>(
         &mut self,
-        src_buf: B,
+        src_buf: Slice<Buf>,
         ctx: &RequestContext,
-    ) -> (B::Buf, Result<(), Error>) {
+    ) -> (Slice<Buf>, Result<(), Error>) {
+        assert_eq!(
+            src_buf.bytes_init(),
+            src_buf.bytes_total(),
+            "caller likely meant to fill the buffer fully"
+        );
```

Comment:
So what does this actually enforce?

`bytes_total` will return `std::cmp::min(Slice::end, buf.len()) - Slice::begin` and `bytes_init` returns `Slice::end - Slice::start`. Hence, we are asserting that `Slice::end <= buf.len()`. Sooo we are checking that when the slice was taken, it only included the initialized part of the IoBuf?

Reply:
> Hence, we are asserting that Slice::end <= buf.len(). Sooo we are checking that when the slice was taken, it only included the initialized part of the IoBuf?

Correct!